Kapow!

jaredhirsch opened the pull request:
Node tests not 100% solid (test automation crew still cranking) but want to get this mainlined sooner than later.

jaredhirsch committed:
kapow! py tests gone

jaredhirsch committed:
copy all code from 6a68/meh and update .gitignore

jaredhirsch committed:
minor renaming to better match unit tests

jaredhirsch committed:
add info to README

jaredhirsch committed:
primary health check working wut wut

jaredhirsch committed:
secondary health check done, code looking fairly clean

jaredhirsch committed:
add test to create secondary user and verify in second browser

jaredhirsch committed:
Extract browser/saucelabs setup to testSetup.

  - still need to extract args from cmd line and inject into testSetup.startup().
  - now you can get the default args by doing browser.newSession(testSetup.sessionOpts)
    or by doing browser.newSession(testSetup.mergeOpts({othercapabilities})
    but it would be even better if you didn't have to remember to do that.
    - OTOH I don't want to hard-code connections between wd and testSetup.
  - added default capabilities to sauce-platforms file

jaredhirsch committed:
update tests to use testSetup, not wd.remote

jaredhirsch committed:
Finish new-user tests, add multiple browsers to testSetup

  - we have primary account creation in all of persona.org, 123done, mfb
  - we have secondary account creation in 123done, mfb, and persona.org,
    including two-browser back-and-forth jumping in mfb and 123done
  TODOs:
    - primary tests seem to break on verification step?
    - figure out why primaryVerify button sticks on first click

jaredhirsch committed:
update other tests to use new testSetup syntax

jaredhirsch committed:
returning-user test: first version done

jaredhirsch committed:
add sauce ID logging to testSetup, fix typo

jaredhirsch committed:
simplify personatestuser API by moving defaults into it

jaredhirsch committed:
WIP public-terminals test; localStorage hack works.

  - localStorage hack works in sauce IE8 and local FF.
  - TODO: now that we have the setup done, figure out which cases to test.

jaredhirsch committed:
add testSetup.setup method for streamlining fixture setup

jaredhirsch committed:
Add test runner info to README. Thx @ozten.

jaredhirsch committed:
Update tests to use testSetup.setup and actually work

jaredhirsch committed:
add sauce test running info to README

jaredhirsch committed:
cleaning up readme markdown

jaredhirsch committed:
final attempt for now at markdown readme cleanup

jaredhirsch committed:
Fix change-passwd test; ensure testSetup waits till personatestusers are ready.

  * Added kriskowal's Q library to synchronize personatestuser waits.
  * Discovered testSetup calls need to be wrapped in an initial vows vow for
    synchronization to happen in cases where personatestuser is sluggish.

Commit:
Make the change-password-test run.

jrgm committed:
- fix the backward order of arguments to wd.prototype.newSession. (We weren't
  actually running any browser but the default FF/Vista one as a result. Now
  that we are trying to run Windows, we don't seem to be able to. But that's
  next to fix.)
- fix some accidental globals and shadows
- use underscore.extend to reduce boilerplate
- remove unused testSetup.mergeOpts method
- node-inspector doesn't like one-liner functions

jrgm committed:
- use lowercase alphanum for restmail email until GH-2551 is fixed
- add back the "click twice" oddity for verifyWithPrimaryButton

Commit:
Starting on the reset password test.

Commit:
If there is no error defined but the response code is not 200, pass the response body as the error.

Commit:
Change the reset-password-test to use restmail.

Commit:
Adding a couple of wrapper functions, not all done yet.

Add lib/user.js with:
* verifyEmail
* getVerifiedUser

jrgm committed:
give the test a name, and default to chrome/vista which does not require a version number

jrgm committed:
only set 'proxy = { proxyType: 'direct' }' for Opera. IE doesn't like it. And add a 'persona' tag to the job

jrgm committed:
add a test to run the frontend /test/ via sauce

jrgm committed:
fix the in-browser javascript for IE8 with _.map()

jrgm committed:
add some mobile devices; update a comment

jrgm committed:
pull out details on which qunit tests failed (if any)

Commit:
After a user resets their password, make sure they can log in to the dialog using the new password.

Commit:
Add remove-email tests

Add a helper to wd-extensions - delay.
Start to standardize CSS selectors.

Commit:
add the removeEmail helper function.

Commit:
Add checks to make sure deleted emails no longer have accounts.

* Add wd.prototype.wclear to enable clearing of functions.

jaredhirsch committed:
Update all tests to synchronize the testSetup call

  The most natural way to do this seemed to be wrapping testSetup.setup() calls
  inside an initial vow. This makes it kinda sloppy if we just want to run
  certain specific tests within a file, but there are other TODOs related to
  enabling that, so it's not breaking anything that was working before.

jaredhirsch committed:
Add runner to ensure no vow keys are ever duplicated

  TODO: when we have a commandline test runner, we'll want to register and
  run tests via that command-line tool, and not inline at the foot of each test

Commit:
Start work on the cancel account test

jaredhirsch committed:
Update tests to use runner, not vowsHarness

Commit:
Adding scripts/run-all.js which allows all the tests to be run.

Commit:
Run the tests through vows instead of through node directly. Report stdout results to a file.

* Add 3 Reporter types, StdOutReporter, StdErrReporter and FileReporter

Commit:
Woot! Add a deploy script which copies the SauceLabs crets to the remote server. Selenium tests are started on Sauce against the ephemeral instance.

Commit:
update the dialog.emailInput and dialog.existingPassword selectors to go along with the changes to remove duplicate IDs

Commit:
Fix the myfavoritebeer URL if using an ephemeral instance.

jrgm committed:
fix the default testname when running under vows, and hoist this out of wd-extensions.js where it didn't belong

jrgm committed:
You can temporarily force a local browser run with `PERSONA_NO_SAUCE`. If you do this, make sure `PERSONA_BROWSER` is set to something that can be run locally.

jrgm committed:
Merge pull request #2653 from jrgm/kapow-stuff

fix testname when running under vows; use PERSONA_NO_SAUCE to run locally

jrgm committed:
add a primary to secondary, mostly stealing work already done

jrgm committed:
a different approach to clicking button#verifyWithPrimary, that saves 20 seconds half the time; also pass done to browser.quit() to avoid a race at shutdown that leaves sauce jobs unfinished

jrgm committed:
Merge pull request #2654 from jrgm/kapow-adding-pri-sec

Kapow adding primary email to account with secondary

jrgm committed:
create an account with primary and add a primary to it

jrgm committed:
Merge pull request #2655 from jrgm/kapow-pri-pri

create an account with primary and add a primary to it

Commit:
Using a new awsbox which supports local scripts to copy sauce credentials to ephemeral instances and start sauce tests.

jrgm committed:
always pass done to browser.quit. Otherwise the sending DELETE session is in a race with process quit, and when that loses, the session lives on for 90 seconds and is marked as error

jrgm committed:
always pass done to browser.quit. Otherwise the sending DELETE session is in a race with process quit, and when that loses, the session lives on for 90 seconds and is marked as error

jrgm committed:
Merge pull request #2657 from jrgm/kapow-browser-quit-then-done

always pass done to browser.quit()

jaredhirsch committed:
to preserve 0.6 node compat, check for existsSync on path and fs

jaredhirsch committed:
to preserve 0.6 node compat, check for existsSync on path and fs

Commit:
Change the awsbox config, localhooks has been renamed to local_hooks. Update awsbox sha in package.json

Commit:
Add the PERSONA_BROWSER environment variable to the sauce.json config file, use this on the remote server to specify which tests should be run.

jaredhirsch committed:
Report Errors correctly by coercing them to string

Commit:
A small cleanup for how output is displayed.

jaredhirsch committed:
Pass asserts into the done callback always.

  There is a weird mismatch in that assert() throws AssertionErrors if an
  assertion fails--yet node works by continuation passing, so that errors
  thrown outside the callback chain aren't handled gracefully.

  In our case, uncaught AssertionErrors cause vows to throw up its hands in
  despair, and not write anything to the XML results file. Since blank files
  aren't valid XML, this causes Jenkins to fail the whole test run on parser
  error grounds. What we want is for it to descriptively fail the build, by
  reading and reporting properly on which tests failed.

  Hopefully confining asserts to the initial error param in the callback chain
  fixes the error reporting headaches.

jaredhirsch committed:
Merge branch 'kapow-assert-inside-callbacks' into kapow

jaredhirsch committed:
make sauce labs tests public by default

lloyd committed:
if .waitFor hits a timeout before event waiting for occurs, invoke complete callback with an error

jaredhirsch committed:
Use local version of asserts which return text, not Errors, on failure

jaredhirsch committed:
break out of chains when an error occurs, using lloyd's wd patch

jaredhirsch committed:
Suppress JSON test output, so that Jenkins gets parseable XML output

Commit:
Upgrade awsbox to 0.3.3

Commit:
Setting up travis to call test_selenium, which starts the test process on testy.

Commit:
exit with a code of 1 if there are is an </error> tag found, 0 otherwise.

Commit:
fix the typo

Commit:
Make sure at least one test suite is run but that there are no errors.

Commit:
Write directly to stderr and stdout instead of using the util functionality. This results in fewer newlines being added to the output.

jaredhirsch commented:
odd that the merge isn't automagic. working on rebasing the branch now.

jaredhirsch committed:
Merge remote-tracking branch 'origin/kapow' into dev

jaredhirsch commented:
ok, that looks better. @lloyd, do your thang :-)

jaredhirsch committed:
Pass extra arg to vows so it reports errors as valid XML.

  For an unknown reason, utils.waitFor causes vows to asplode when element
  lookup times out--vows reports these failures using the console reporter,
  instead of the xunit reporter. FWIW the relevant snippet seems to be
  https://github.com/cloudhead/vows/blob/38817c1b2670db45e982b52b30ae044e6c683130/lib/vows.js#L204

  To work around this, we ask vows to spawn each test in a separate process,
  via the "-i" option, and it reports failures correctly as well-formed XML.

jrgm commented on .gitignore:
```diff
@@ -16,11 +16,6 @@
-automation-tests/persona_server/results/*
-/automation-tests/chromedriver.log
-/.coverage_data
-/cover_html
```
Those last two `.coverage_data` and `cover_html` were added to `dev` in PR-2272 and should not be removed. But if it's a hassle to rebase again, don't worry; they'll eventually get added back in some other PR.

jaredhirsch commented:
hell yes. vows finally giving jenkins something to work with. couple of broken tests, but breakage now leads to well-formed XML. we'll get to fixing the tests next :-)

lloyd committed:
document why I believe we need a .wfind() before the .wclick() in add-primary-to-primary

lloyd commented:
ok, debating whether landing in dev _right now_ is the right thing to do.  The only pre-requisites I would consider hard are:
1. an easy way to run all tests all browsers
2. test stability (90% of the time all tests all browsers succeeds)
3. jenkins integration (daily?)

This is a subset of what is mentioned in the planning etherpad. 

once 1 is done, I'd love to contribute to 2 by using it.

all that said, is there a good reason to merge into dev before these 3 top level requirements are met?

/cc @shane-tomlinson @jrgm

lloyd commented:
yeah, and if 3 got done and we were running tests hourly using jenkins, then as soon as we have 50 runs with < 5 failures, we know we've attained 2...

lloyd committed:
update vows harness to support .bailOnError behavior by default.  This allows us to get better error information out of vows, as well as not run tests that depend on the success of the former.  BONUS: a .cleanup can be specified that will always be run.  useful for shutting down saucelabs sessions rather than letting them take an additional 90s to timeout

lloyd committed:
upgrade to new version of wd that fixes a bug whereby .onError handlers were not getting invoked, and improves semantics of automatic error detection so that we can ignore errors in some cases.  also, allow programmatic halting of queued chain functions

lloyd committed:
whitespace change, consistent indentation

lloyd committed:
for returning-user tests, quit in a cleanup handler so that we always cancel the session promptly, even when it errors (means 90s less wait time when you're trying to debug a test by looking a the protocol messages that were exchanged via sauce)

lloyd committed:
fix returning-user test - allow clicking of 'not my computer' to fail, given that this button may not be displayed if 60s have not elapsed since account verification

shane-tomlinson commented:
@lloyd - I agree with your requirements with the following caveats.

For #1, we have to define which browsers make this list and with what frequency tests are run against all browsers. Is this on checkin/push to origin, on merge, etc? What happens for developers that first push to their own repos and then submit a PR, do these checkins automatically kick off the full test set? Currently we have no way of telling travis "run Selenium tests from this repo, but not from that repo over there."

For #3, Jenkins integration against which branch - dev, stage, prod or against any old branch?

Commit:
Add the ability to run all tests against all supported SAUCE browsers by specifying PERSONA_BROWSER=all when running run-all.js

Commit:
Get the repo URL from the git repo and send that along with the request. Helps establish some sort of ACL on testy-listener so that it will only run tests on repos from whitelisted users.

Commit:
Up the number of concurrent runners to 30.

Commit:
Add the ability to specify the REPO_URL and REPO_SHA on the command line when calling run-all.js. Up the number of concurrent test runners to 30.

Commit:
Fix getOrigin returning an origin with a \n appended to it.

Commit:
Hook up to the process "exit" message instead of "end" message.

This fixes the problem where the results were never actually checked and travis thought all tests were passing.

jaredhirsch commented:
So, the tests look green on Jenkins with all-browsers selected: https://ci.mozilla.org/job/temporary-test-job/236/console . I'll put that job on a short cron to get some stability numbers, and update this ticket tomorrow/late tonight with details. [update--I got the env var wrong and was only running against chrome-vista, doh. more info soon.]

@shane-tomlinson, to address some of your questions:

**Which browsers should we use?** The browsers we worried about with our selenium tests in the past were the ones in the sauce-browsers list, excluding the mobile platforms @jrgm added, but those don't track precisely to our official list of supported browsers/platforms (see [MDN](https://developer.mozilla.org/en-US/docs/Persona/Browser_compatibility)). I'd say, running that set of browsers (those in sauce-browsers excluding mobile) is a good starting point.

Code proposal: we generalize the current browser info file, and pass the file name optionally on the command line. Since sauce-platforms is just an exported object, we can have a "core" file that's required by an "extras" list, to hold stuff like mobile browsers that aren't ready for primetime yet. (I'll copy this idea into the test-automation-spec etherpad so it doesn't get lost). This is a nice-to-have.

**Which branches do we run against, and how often?** This is mentioned at a high level in the [automated testing wiki page on github](https://github.com/mozilla/browserid/wiki/Automated-Browser-Testing). 
- We'll run all tests/all browsers against stage and prod on a long cron (every 4 or 6 hours). 
- We'll run just the health-check tests, on just one stable browser/OS combo, against stage and prod every 5 minutes, as a monitoring stopgap. (This is a temporary monitor and I'll be improving our monitoring after this crew spins down.) 
- We'll run all tests/all browsers against dev on git push, as well as on a 4 or 6 hour cron. 

This is all already implemented in Jenkins on ci.mozilla.org; I just need to update the invocations to use our node tests instead of the old py tests.

**What about Travis testing PRs where the branch lives on somebody's fork?** I dunno. That's a good question. I guess Travis needs to give the git repo URL along with the SHA? Is that a tenable solution? Otherwise, maybe we can amend our git workflow practices to submit PRs from a branch living on mozilla/browserid.

jaredhirsch committed:
Temporarily remove mobile platforms from saucelabs selenium routine

Commit:
Use libraries to create directories, copy objects, copy environment.

* use mkdirp, toolbelt used in testy.
* Correctly set the number of runners in run-all.js from the RUNNERS environment variable.

shane-tomlinson commented:
@6a68:

> What about Travis testing PRs where the branch lives on somebody's fork? I dunno. That's a good question. I guess Travis needs to give the git repo URL along with the SHA? Is that a tenable solution? Otherwise, maybe we can amend our git workflow practices to submit PRs from a branch living on mozilla/browserid.

This is what I suspected, so I did that yesterday. ;) What I am trying to get at is an ACL of some sort. What I have done as a super basic measure is to create a whitelist set of repo owners whose repo's can be tested via testy. A POST request to testy passes the git repo URL as well as the sha. Testy looks at the repo URL, divines the repo owner, checks against the whitelist and proceeds.

This is a start, but it is still open to abuse as any random yahoo could POST to testy the URL of one of the approved owner's repos and kick off sauce tests.

What I would like is to whitelist a set of IP addresses so that we can check the ip address of the requester, and whitelist Travis IP addresses only or use HTTP authentication with a secret stored on travis to sign in. Dunno. The former is not yet possible as the awsbox proxy does not forward the ip address of the requester from the proxy to the main server. The latter I thought the whole reason we were doing this exercise is so we didn't have to store secrets on Travis.

Commit:
Exit with the correct status code, 0 if success, 1 if failure.

Commit:
Use toolbelt's copyExtendEnv function when extending the environment. Fix jshint errors.

Commit:
Fix up the error checker when looking at the Sauce results.

jaredhirsch committed:
minor edit: remove stale todo

Commit:
Capture file open/write/close errors so that it does not bring the entire process down.

Commit:
Add mkdirp to the list of dependencies.

Commit:
Only enable vista_chrome for now until we can better stablize the tests.

jaredhirsch committed:
Use assert.fail in vows_harness, so XML will show failure/error.

  If we just throw a string, we get status 'skipped', which Jenkins doesn't
  interpret as a failure.

jaredhirsch committed:
Provide more context in the runner stderr output

jrgm committed:
Optionally add tags to sauce sessions from the environment

jrgm committed:
use done as the callback for browser.quit

jaredhirsch committed:
Pass test file name to 'classname' vows xml attribute

jaredhirsch committed:
Minor edit: add missing 'path' include to frontend-qunit-tests

jaredhirsch committed:
Minor edit: screwing around with Jenkins readability of stderr

Commit:
Add wd.prototype.wclickIfExists which waits one second and clicks an element if it exists. If not, no problem.

Use this in remove-email.js where the verifyWithPrimaryButton sometimes needs clicked twice and the "Is this your computer" can appear if the test takes longer than one minute.

Commit:
Use wclickIfExists in place of the timeout solution when the "Is this your computer" screen is shown.

Commit:
Change the ordering of the tests so that secondary tests are run before primary tests. This masks a timing problem that keeps occurring when trying to enter a secondary address and the password fields are not found.

Commit:
Add wd-extensions.prototype.wgetAttribute, use that in the returning-user tests to check whether a radio button is selected.

Commit:
Add a cleanup function so browsers are always torn down.

Commit:
Update the timeout error message to indicate which element search timed out.

Commit:
Adding info about how to run all tests in parallel.

Also add info on wclickIfExists, wgetAttribute, wclear and delay

lloyd committed:
don't run saucelabs tests at push time

lloyd committed:
use optimist for command line parsing.  implement --help and --platform (backwards compatible)

lloyd committed:
(automated testing) implement --list-tests

lloyd committed:
(automation tests) implement --list-platforms

lloyd committed:
(automation tests) implement --parallel with a default value of 10 and defer to environment for back compat

jaredhirsch committed:
re-enable non-mobile sauce platforms for stability stats gathering

jaredhirsch committed:
Update browsers: ff13->16 on linux, win; add ie10

lloyd committed:
(automation tests) allow platform to be specified using globs, improve reporting for summary of what will be done

lloyd committed:
allow persona environment to be specified on the command line, back compat with ENV

lloyd committed:
allow a glob to be specified to match individual tests

lloyd committed:
better -h for --tests

lloyd committed:
initial cleanup of test output

lloyd committed:
make run-all.js friendly for local testing, add --local flag

lloyd committed:
merge improvements to run-all.js to make it nicer to invoke from the command line

Conflicts:
	automation-tests/lib/sauce-platforms.js

lloyd committed:
move minimatch down into automation-tests, that's the only place it is used and in the jenkins env we don't npm install the lower level

jaredhirsch commented:
OK, so we've got 90%+ stability on all platforms except those added really recently (mac safari 5, win 8 IE10, mobile).

Suggest merging in, stabilizing those new platforms separately, and merging those fixes later.

Thoughts?

lloyd commented:
dig it, will do a pass in the morning and merge!

On Nov 7, 2012, at 7:41 PM, Jared Hirsch notifications@github.com wrote:

> OK, so we've got 90%+ stability on all platforms except those added really recently (mac safari 5, win 8 IE10, mobile).
> 
> Suggest merging in, stabilizing those new platforms separately, and merging those fixes later.
> 
> Thoughts?
> 
> —
> Reply to this email directly or view it on GitHub.

Commit:
Bug fixin' & cleanup of run-all.js and test-finder.js. @lloyd did some great work to make run-all.js easier to use, I am furthering his efforts.

* Re-enable platforms when calling Selenium.
* Start to split out functions.
* Replace _ in filenames with - for consistency.
* Show the platform in the stderr output.

Commit:
Update the documentation for lloyd's command line changes to run-all.js

jaredhirsch committed:
Tweak sauce settings: add 3-min global timeout, 30-sec idle timeout

Commit:
Add the ability to specify which vows reporter to use from the command line

Commit:
Only open files when actually writing to them. Helps OSX out when running massively parallel tests.

lloyd committed:
fix bug where test glob not passed into finder

lloyd committed:
upgrade to a new patched version of wd that handles ECONNRESET gracefully and allows the client to explicitly fiddle the http inactivity timeout

lloyd committed:
work around aw snap in remove-email test issue #2691

lloyd committed:
as a start, report non-zero exit codes on the console

lloyd committed:
make the world safe for 'any' for awesome local testing

lloyd committed:
Revert "Add the ability to specify which vows reporter to use from the command line"

This reverts commit b5859b24d3fbd6af8d6763f91fe43453ad490976.

jaredhirsch committed:
lengthen idle-timeout so dual-session tests don't fail

lloyd committed:
more friendly console reporting for run-all.js

lloyd committed:
tighten up help language

lloyd committed:
fix bug where tests not started in the first batch would not get included in aggregated results

lloyd committed:
add cleanup of sauce VMs to health check tests so they don't hang for 90 seconds in case of failure

lloyd committed:
fix cleanups so that they run properly under the vows harness, add more cleanups to more tests so we don't leave vms lying around to timeout.  also, fix proper exit codes under xunit and console output.

lloyd committed:
remove debug message

lloyd committed:
the process must wait for cleanups to complete, otherwise they don't actually get to make their HTTP transactions and cleanup VMs for failed tests

jaredhirsch committed:
Improve setup and teardown across all tests

  - update setup and teardown in all tests
  - testSetup now really handles fixture lifecycle:
    - init browser sessions using testSetup
    - teardown browsers using testSetup
  - use Q to handle control flow for async multi-browser teardown
    - quitIfDefined caused barfing, passing undefined args into a function call

jaredhirsch committed:
rearrange test-setup to more clearly separate private/public functions

jaredhirsch committed:
bah, missed one stray newSession call.

jrgm committed:
tag test with the environment; drop the redundant persona which was useful in shared environment

jrgm committed:
repair this test; gets 10/10 on vista_{chrome,firefox_16}, although this test creates a lot of sessions

Commit:
Update post-update.js so that it exits with the same exit code as run-all.js. Depend on this exit code making back to testy eventually so that test_selenium can exit with the same exit code and travis correctly marks the tests as pass/fail.

Commit:
Only enable Chrome and disable almost all tests until we have better test stability.

Commit:
Use a GET request when requesting selenium results.

jaredhirsch committed:
Re-enabling tests that worked recently & switching active platform to FF/vista

jaredhirsch committed:
Whoops, disabled the wrong test.

jaredhirsch committed:
Rework double button click in primary dialog.

  I can't repro locally, but sauce's vista/FF 16 consistently seems to
  succeed with the first click, then sadly wait forever and never make
  the second click. Trying to replace the second click with a click-if-exists.
  Hopefully this works on the sauce side.

jaredhirsch committed:
Revert "work around aw snap in remove-email test issue #2691"

  This reverts commit 6a6cf59.

  The about:blank fix causes rampant destruction on FF and Chrome , reverting
  for the moment. We'll find a different fix when we focus on Chrome next week.

Commit:
Move sauce-platforms to the config directory.

Commit:
Clean up the JSHint errors and warnings.

Commit:
Make sure all browser.chain() invocations pass an onError callback.

Commit:
Keep track of the current window so that it is possible to check whether the last opened window is still open. This prevents Firefox from hanging when checking for an element in a window that has already been closed.

When checking if an element exists, wait for 15 seconds before the check to ensure that the dialog goes away if it is going to go away.

Commit:
Add wclickIfExists on the eyedee.me create account button for IE.

jaredhirsch committed:
Reorder vows in new-user-secondary. Fixes #2704.

  When I swapped the browser.newSession calls for testSetup.newBrowserSession
  calls, the vows got split in the wrong order, so that null was getting passed
  as the link to the browser.get call.

jaredhirsch commented:
zomg so many commits! looking stable on FF 16/Vista, what else needs to happen to merge this puppy in today?

jaredhirsch commented:
r+ from lloyd in IRC. merging.
